### DISCUSSION TITLE
How to format members Roles expressions?

### DISCUSSION THREAD

**Discussion by rvgfox** [TabularEditor/TabularEditor3#1205] [🙋♀️TE3 Q&A] [answered]
hi,

I'm using in my scripts:

FormatDax(Model.AllMeasures, true);

But' obviously it doesn't works with the member's role expressions.

Is it possible to do that?

**Comment by DBojsen** [comment, marked as answer]
Hi @rvgfox 

This should give you what you need:
``` c#
List<string> messages = new List<string>();

// Run through all roles
foreach (var role in Model.Roles)
{
    // Run through all tables to see if an RLS expression is set
    foreach (var table in role.RowLevelSecurity.Keys)
    {
        if (!string.IsNullOrEmpty(role.RowLevelSecurity[table]))
        {
            // Format the DAX expression
            role.RowLevelSecurity[table] = FormatDax(role.RowLevelSecurity[table]);
            messages.Add($"Role {role.Name} - Table {table}");
        }
    }
}

if (messages.Count > 0)
{
    Info("Formatted role expressions for:\r\n" + string.Join("\r\n", messages));
}
else
{
    Info("Didn't find any role expressions to format!");
}
```

Best Regards

David - TE3 Support

**Reply by rvgfox** [reply]
@DBojsen I'm trying to run that script in TE2 but I get an error;

![image](https://github.com/TabularEditor/TabularEditor3/assets/72302636/8065cc3b-a9cf-4a10-9888-e0be9e94dc0f)

How can I solve it?

Thanks in advance

**Comment by rvgfox** [comment]
Thank you very much @DBojsen it works perfectly.

**Comment by DBojsen** [comment]
Hi @rvgfox 

Here is a better script that works in both TE3 & TE2:
``` C#
List<string> messages = new List<string>();

// Run through all roles
foreach (var role in Model.Roles)
{
    // Run through all table permissions
    foreach (var permission in role.TablePermissions)
    {
        FormatDax(permission);
        messages.Add("Role " + role.Name + " - Table " + permission.TableName);
    }
}

if (messages.Count > 0)
{
    Info("Formatted role expressions for:\r\n" + string.Join("\r\n", messages));
}
else
{
    Info("Didn't find any role expressions to format!");
}
```

**Reply by DBojsen** [reply]
If you don't care about the info box at the end, you can do it all in one line:
``` c#
FormatDax(Model.Roles.SelectMany(r => r.TablePermissions));
```

**Comment by rvgfox** [comment]
Thank you very much @DBojsen 

I've tested:

`FormatDax(Model.Roles.SelectMany(r => r.TablePermissions));`

And now, in TE2 I haven't the error message, thanks.

But I'm testing with this measure:

`Calculate(
  'Facturas Ventas'[Kilos Cafe Facturas Ventas]  ,
    Filter( 'Facturas Ventas' ,
'Facturas Ventas'[Id_sk_Contratos]  <> -1)
)`

With TE3 it's formated to:

`CALCULATE( [Kilos Cafe Facturas Ventas], FILTER( 'Facturas Ventas', 'Facturas Ventas'[Id_sk_Contratos] <> -1 ) )`

It's correct acording with BPA, but with TE2 it's formated to:

`CALCULATE (
    'Facturas Ventas'[Kilos Cafe Facturas Ventas],
    FILTER (
        'Facturas Ventas',
        'Facturas Ventas'[Id_sk_Contratos] <> -1
    )
)`

It doesn't follow the BP because "Measure references should not be qualified with the table name"

Some ideas?

**Comment by DBojsen** [comment]
Hi @rvgfox 

TE3 uses our own formatting engine by default.
TE2 uses daxformatter.com which doesn't have the context of your model.

**Comment by rvgfox** [comment]
Thanks @DBojsen Will them use both, TE2 and TE3, the same engine in a near future?

**Comment by DBojsen** [comment]
Hi @rvgfox 

Sorry - your comment slipped through my holiday email madness.
Both can use daxformatter.com, but there are no plans to bring our own formatting engine to TE2.

/ David

**Reply by mlonsk** [reply]
Hi @rvgfox 
Just to follow up on David's remark, you can choose which formatter you want to use in Tabular Editor 3 Preferences: 

![image](https://github.com/TabularEditor/TabularEditor3/assets/11939114/e982d542-b449-4b88-b89f-df89d3c09760)

Just in case you want the same output in TE2 and TE3.
